File: output/gpt4/js/Solution19_gpt.js
class BinaryIndexedTree {
    constructor(n) {
        this.tree = new Array(n + 1).fill(0);
    }

    update(index, value) {
        while (index < this.tree.length) {
            this.tree[index] += value;
            index += index & -index;
        }
    }

    query(index) {
        let sum = 0;
        while (index > 0) {
            sum += this.tree[index];
            index -= index & -index;
        }
        return sum;
    }
}

var countSmaller = function(nums) {
    const sortedNums = [...nums].sort((a, b) => a - b);
    const indices = new Map();
    for (let i = 0; i < nums.length; i++) {
        indices.set(sortedNums[i], i + 1);
    }

    const result = [];
    const bit = new BinaryIndexedTree(nums.length);
    for (let i = nums.length - 1; i >= 0; i--) {
        const index = indices.get(nums[i]);
        result.push(bit.query(index - 1));
        bit.update(index, 1);
    }
    return result.reverse();
};

File: output/gpt4/java/Solution19_gpt.java
import java.util.*;

class Solution19_gpt {
    public List<Integer> countSmaller(int[] nums) {
        int[] sortedNums = nums.clone();
        Arrays.sort(sortedNums);
        Map<Integer, Integer> indices = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            indices.put(sortedNums[i], i + 1);
        }
        
        List<Integer> result = new ArrayList<>();
        BinaryIndexedTree bit = new BinaryIndexedTree(nums.length);
        for (int i = nums.length - 1; i >= 0; i--) {
            int index = indices.get(nums[i]);
            result.add(bit.query(index - 1));
            bit.update(index, 1);
        }
        Collections.reverse(result);
        return result;
    }

    private class BinaryIndexedTree {
        int[] tree;

        public BinaryIndexedTree(int n) {
            tree = new int[n + 1];
        }

        public void update(int index, int value) {
            while (index < tree.length) {
                tree[index] += value;
                index += index & -index;
            }
        }

        public int query(int index) {
            int sum = 0;
            while (index > 0) {
                sum += tree[index];
                index -= index & -index;
            }
            return sum;
        }
    }
}
